### PR TITLE
Et/v2 register explain hook

### PR DESCRIPTION
**Pull request by vinnamkim**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by negvet — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]
Overall seems nice, but I have a concern that user might confuse between callbacks and hooks. I don't have nice suggestion, but let's think about more proper naming for these classes.

**Comment by vinnamkim** [comment]
> Overall seems nice, but I have a concern that user might confuse between callbacks and hooks. I don't have nice suggestion, but let's think about more proper naming for these classes.

I just wanted to keep the momentum from OTX 1.* which has been used a PyTorch Module hook for XAI. However, I think that it would be better to expose a dedicated `forward` function for XAI functionality. For example,

```python
class ExplainableOTXClsModel:
     def forward_with_xai(self, inputs: OTXBatchDataEntity) -> OTXBatchPredWithXAI:
          ...
          return prediction_with_xai
```

Could we move forward to this way in the future? @negvet

**Review by harimkang — APPROVED** [review]

**Comment by negvet** [comment]
> > Overall seems nice, but I have a concern that user might confuse between callbacks and hooks. I don't have nice suggestion, but let's think about more proper naming for these classes.
> 
> I just wanted to keep the momentum from OTX 1.* which has been used a PyTorch Module hook for XAI. However, I think that it would be better to expose a dedicated `forward` function for XAI functionality. For example,
> 
> ```python
> class ExplainableOTXClsModel:
>      def forward_with_xai(self, inputs: OTXBatchDataEntity) -> OTXBatchPredWithXAI:
>           ...
>           return prediction_with_xai
> ```
> 
> Could we move forward to this way in the future? @negvet

Yeah, we will look into this direction

File: src/otx/core/model/entity/classification.py [hunk 1]
```diff
@@ -29,9 +30,52 @@
     from torch import device, nn
 
 
-class OTXMulticlassClsModel(
-    OTXModel[MulticlassClsBatchDataEntity, MulticlassClsBatchPredEntity],
-):
+class ExplainableOTXClsModel(OTXModel[T_OTXBatchDataEntity, T_OTXBatchPredEntity]):
```
Comment (samet-akcay) [inline src/otx/core/model/entity/classification.py:33]: Isn't it possible to attach the hook to an existing classification model? I was wondering if it is really needed to create an additional object for this? I have a concern that there is a growing list of tasks/objects we have, which will ultimately be getting a bit confusing to users.

**Pull request merged**